allennlp/models/graph_parser.py

```diff
@@ -0,0 +1,341 @@
+from typing import Dict, Optional, Tuple, Any, List
+import logging
+import copy
+
+from overrides import overrides
+import torch
+from torch.nn.modules import Dropout
+import numpy
+
+from allennlp.common.checks import check_dimensions_match, ConfigurationError
+from allennlp.data import Vocabulary
+from allennlp.modules import Seq2SeqEncoder, TextFieldEmbedder, Embedding, InputVariationalDropout
+from allennlp.modules.matrix_attention.bilinear_matrix_attention import BilinearMatrixAttention
+from allennlp.modules import FeedForward
+from allennlp.models.model import Model
+from allennlp.nn import InitializerApplicator, RegularizerApplicator, Activation
+from allennlp.nn.util import get_text_field_mask
+from allennlp.nn.util import get_lengths_from_binary_sequence_mask
+from allennlp.training.metrics import F1Measure
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Model.register("graph_parser")
+class GraphParser(Model):
+    """
+    A Parser for arbitrary graph stuctures.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``, required
+        Used to embed the ``tokens`` ``TextField`` we get as input to the model.
+    encoder : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use to generate representations
+        of tokens.
+    tag_representation_dim : ``int``, required.
+        The dimension of the MLPs used for arc tag prediction.
+    arc_representation_dim : ``int``, required.
+        The dimension of the MLPs used for arc prediction.
+    tag_feedforward : ``FeedForward``, optional, (default = None).
+        The feedforward network used to produce tag representations.
+        By default, a 1 layer feedforward network with an elu activation is used.
+    arc_feedforward : ``FeedForward``, optional, (default = None).
+        The feedforward network used to produce arc representations.
+        By default, a 1 layer feedforward network with an elu activation is used.
+    pos_tag_embedding : ``Embedding``, optional.
+        Used to embed the ``pos_tags`` ``SequenceLabelField`` we get as input to the model.
+    dropout : ``float``, optional, (default = 0.0)
+        The variational dropout applied to the output of the encoder and MLP layers.
+    input_dropout : ``float``, optional, (default = 0.0)
+        The dropout applied to the embedded text input.
+    edge_prediction_threshold : ``int``, optional (default = 0.5)
+        The probability at which to consider a scored edge to be 'present'
+        in the decoded graph. Must be between 0 and 1.
```

Comment:
You might note that this is only used when making predictions, and not during training.